contracts/gas/GasEstimate.sol

```diff
@@ -0,0 +1,143 @@
+ */
+contract GasEstimate is IGasEstimate {
+    // keccak256('GasEstimate.Slot') - 1
+    bytes32 internal constant GAS_SERVICE_SLOT = 0x2fa150da4c9f4c3a28593398c65313dd42f63d0530ec6db4a2b46e6d837a3902;
+
+    struct GasServiceStorage {
+        mapping(string => GasInfo) gasPrices;
+    }
+
+    /**
+     * @notice Sets the gas price for a specific chain.
+     * @dev This function is called by the gas oracle to update the gas price for a specific chain.
+     * @param chain The name of the chain
+     * @param gasInfo The gas info for the chain
+     */
+    function _setGasInfo(string calldata chain, GasInfo calldata gasInfo) internal {
+        emit GasInfoUpdated(chain, gasInfo);
+
+        _gasServiceStorage().gasPrices[chain] = gasInfo;
+    }
+
+    /**
+     * @notice Estimates the gas fee for a contract call on a destination chain.
+     * @param destinationChain Axelar registered name of the destination chain
+     * param destinationAddress Destination contract address being called
+     * @param executionGasLimit The gas limit to be used for the destination contract execution,
+     *        e.g. pass in 200k if your app consumes needs upto 200k for this contract call
+     * @return gasEstimate The cross-chain gas estimate, in terms of source chain's native gas token that should be forwarded to the gas service.
+     */
+    function estimateGasFee(
+        string calldata destinationChain,
+        string calldata, /* destinationAddress */
+        bytes calldata payload,
+        uint256 executionGasLimit
+    ) external view returns (uint256 gasEstimate) {
+        GasServiceStorage storage slot = _gasServiceStorage();
+        GasInfo storage gasInfo = slot.gasPrices[destinationChain];
+
+        gasEstimate = gasInfo.baseFee + (executionGasLimit * gasInfo.relativeGasPrice);
+
+        // if chain is L2, compute L1 data fee using L1 gas price info
+        if (gasInfo.l1ToL2BaseFee != 0) {
+            gasEstimate += computeL1ToL2Fee(
+                destinationChain,
+                payload,
+                slot.gasPrices['ethereum'].relativeGasPrice,
+                gasInfo.l1ToL2BaseFee
+            );
+        }
+    }
+
+    /**
+     * @notice Computes the L1 to L2 fee for a contract call on a destination chain.
+     * @param destinationChain Axelar registered name of the destination chain
+     * @param payload The payload of the contract call
+     *  param l1GasPrice The gas price on the source chain
+     * @param l1ToL2BaseFee The base fee for L1 to L2
+     * @return l1DataFee The L1 to L2 data fee
+     */
+    function computeL1ToL2Fee(
+        string calldata destinationChain,
+        bytes calldata payload,
+        uint256, /* l1GasPrice */
+        uint256 l1ToL2BaseFee
+    ) internal pure returns (uint256) {
+        if (keccak256(bytes(destinationChain)) == keccak256(bytes('optimism'))) {
```

Comment:
let's use the L2 field in GasInfo as an enum instead that dictates the L2 pricing to use instead of hardcoded chain names (that are customizable, for e.g. on testnet we use `optimism-sepolia` for the name)